src/Gaffer/ValuePlug.cpp

```diff
@@ -1142,21 +1215,16 @@ void ValuePlug::dirty()
 	// The invalid entries will be evicted by the LRU rules in due
 	// course.
 	m_dirtyCount++;
-	if(
 		// If our count has wrapped back to 0, it's possible (but
 		// incredibly unlikely) that we could match a stale value
 		// in the cache, so we must do a brute force clear.
-		m_dirtyCount == 0 ||
-		// If a node has bugs in its `DependencyNode::affects()`
-		// implementation, it could also end up using stale cache entries.
-		// We provide the GAFFER_CLEAR_HASHCACHE_ON_DIRTY environment
-		// variable to allow a temporary fallback to the old behaviour
-		// while the bugs are fixed.
-		g_clearHashCacheOnDirty
-	)
+	if( m_dirtyCount >= DIRTY_COUNT_RANGE_HALF )
```

Comment:
I think I mentioned this before, but there is existing code in `~ValuePlug` that also needs to be updated to wrap at `DIRTY_COUNT_RANGE_HALF` rather than `numeric_limits<uint64_t::max()>`.

Reply:
I've taken a swing at this in a82db4b, though it was a bit more complicated than I expected.

As far as I can tell, the previous version of this code would completely fail most of the time.  For simplicity, assume max_uint == 1000.  You have some plug with a dirty count of 998.  Then it gets dirtied 4 times, and then deleted.  It now has a m_dirtyCount of 2, and would not have triggered the previous special case branch.  The dirty propogation that pushes it to wrap would have done a clearCache anyway, so that might be OK, but the value of g_dirtyCountEpoch might still be set at 900 or something, due to the `std::max`.  This would appear to permanently disable g_dirtyCountEpoch ( with a chance to reenable if you happen to land exactly on max_uint the next time it wraps around ).

The version I've now written I think is correct at least, though it does seem like a substantial performance hazard.  I hadn't really thought through how bad the rollover scenario could get, particularly because it's plausible to have many plugs with similar dirty counts - it's not one global clearCache, it's one per plug ... and with this new code, likely 2 per plug.

I thought about fixing that extra clear per plug, and I think it would involve:
* only clear when you land exactly on the wraparound, like before
* in cases where you have wrapped, but not cleared, you need to do a std::max with the current g_dirtyCountEpoch, but ignoring values that haven't wrapped yet ( just ignoring values > DIRTY_COUNT_RANGE_HALF / 2 would do the trick )

That would increase the complexity of this code though, so I've currently just done the simplest thing that I think should give accurate results.